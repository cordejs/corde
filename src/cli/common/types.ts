import { Command } from "commander";

export type ParamsFrom = "args" | "options" | "both";

export interface ICommandOptions {
  flags: string;
  description?: string;
  defaultValue?: string;
}

export interface ICliCommandConstructor {
  name?: string;
  program: Command;
  paramsFrom?: ParamsFrom;
}

export interface ICliCommand {
  readonly paramsFrom?: ParamsFrom;
  readonly command: Command;
  handler(...args: any[]): Promise<void> | void;
}

export interface ICliCommandConstructable {
  new (command: Command): ICliCommand;
}
