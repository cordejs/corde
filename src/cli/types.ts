export interface ICommand {
  action(...args: any[]): Promise<void> | void;
}
