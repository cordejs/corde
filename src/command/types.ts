import { ICordeBot } from "../types";

export interface ICreateMatcherParam {
  matcher?: string;
  isNot?: boolean;
  commandName?: string | boolean | number;
  channelId?: string;
  guildId?: string;
  isDebug?: boolean;
  isCascade?: boolean;
  cordeBot?: ICordeBot;
  trace?: string;
  utility?: boolean;
  mustSendCommand?: boolean;
}
