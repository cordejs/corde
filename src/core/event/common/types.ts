import { Client, ClientEvents } from "discord.js";

export type EventNames = keyof ClientEvents;

export type Spread<T extends EventNames> = ClientEvents[T][1] extends undefined
  ? ClientEvents[T] extends (infer U)[]
    ? U
    : never
  : ClientEvents[T];

export interface IDiscordEvent<TEvent extends EventNames, TFilter> {
  canListen(): boolean;
  getIntents(): string[];
  on(fn: (...args: ClientEvents[TEvent]) => void): this;
  once(filter?: TFilter): Promise<Spread<TEvent>>;
}

export interface IDiscordEventConstructable<TEvent extends EventNames, TFilter> {
  new (client: Client): IDiscordEvent<TEvent, TFilter>;
}

export type IntentType =
  | "GUILDS"
  | "GUILD_MEMBERS"
  | "GUILD_BANS"
  | "GUILD_EMOJIS_AND_STICKERS"
  | "GUILD_INTEGRATIONS"
  | "GUILD_WEBHOOKS"
  | "GUILD_INVITES"
  | "GUILD_VOICE_STATES"
  | "GUILD_PRESENCES"
  | "GUILD_MESSAGES"
  | "GUILD_MESSAGE_REACTIONS"
  | "GUILD_MESSAGE_TYPING"
  | "DIRECT_MESSAGES"
  | "DIRECT_MESSAGE_REACTIONS"
  | "DIRECT_MESSAGE_TYPING"
  | "GUILD_SCHEDULED_EVENTS";

export type EventClass = {
  [k in IntentType]: {
    bitfield: number;
    events: EventNames[];
  };
};
