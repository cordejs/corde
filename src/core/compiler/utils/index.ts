export * from "./import-js-in-directory";
export * from "./was-file-modified";
