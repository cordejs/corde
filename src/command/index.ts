/* eslint-disable no-console */
import { forEachPropsName } from "../utils/forEachPropsName";
import { getStackTrace } from "../utils/getStackTrace";
import { isNullOrUndefined } from "../utils/isNullOrUndefined";
import { any } from "../expect/asymmetricMatcher";
import * as matchers from "./matches";
import runtime from "../core/runtime";
import { ICordeBot, ITestReport } from "../types";
import { CommandState } from "./matches/CommandState";

interface ICreateMatcherParam {
  matcher: string;
  isNot: boolean;
  commandName?: string | boolean | number;
  channelId?: string;
  guildId?: string;
  isDebug: boolean;
  isCascade?: boolean;
  cordeBot?: ICordeBot;
  trace?: string;
  mustSendCommand?: boolean;
}

interface IMatcher {
  (props: CommandState, ...args: any[]): Promise<ITestReport>;
}

type KeyOfMatcher = keyof typeof matchers;

function pickFn(name: KeyOfMatcher) {
  return matchers[name] as any as IMatcher;
}

function buildAndMatcherFunctions(params: ICreateMatcherParam) {
  const { commandName, isDebug, channelId, cordeBot, mustSendCommand } = params;

  const commandReturn: any = {};
  commandReturn.not = {};

  forEachPropsName(matchers, (matcher) => {
    commandReturn[matcher] = createMatcherFn({
      commandName,
      isDebug,
      isNot: false,
      matcher,
      channelId,
      cordeBot,
      mustSendCommand,
    });
    commandReturn.not[matcher] = createMatcherFn({
      commandName,
      isDebug,
      isNot: true,
      matcher,
      channelId,
      cordeBot,
      mustSendCommand,
    });
  });

  return commandReturn;
}

class CommandPromise extends Promise<any> implements corde.ICommandPromise {
  private readonly _param: ICreateMatcherParam;

  get and(): corde.AllCommandMatches {
    return buildAndMatcherFunctions({
      mustSendCommand: false,
      ...this._param,
    });
  }

  constructor(
    executor: (resolve: (value: any) => void, reject: (reason?: any) => void) => void,
    param: ICreateMatcherParam,
  ) {
    super(executor);
    this._param = param;
  }

  catch<TResult = never>(
    onrejected?: ((reason: any) => TResult | PromiseLike<TResult>) | null,
  ): CommandPromise {
    super.catch(onrejected);
    return this;
  }

  finally(onfinally?: (() => void) | null): CommandPromise {
    super.finally(onfinally);
    return this;
  }

  then<TResult1 = any, TResult2 = never>(
    onfulfilled?: ((value: any) => TResult1 | PromiseLike<TResult1>) | null,
    onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | null,
  ): CommandPromise {
    super.then(onfulfilled, onrejected);
    return this;
  }

  [Symbol.toStringTag]: string;
}

async function resolveTestFunction(
  params: ICreateMatcherParam,
  fn: (...args: any[]) => Promise<ITestReport>,
): Promise<any> {
  try {
    const report = await fn();
    if (!report.pass && !params.isDebug) {
      report.trace = params.trace;
    }

    runtime.internalEvents.emit("test_end", report);

    if (params.isDebug) {
      return report;
    }
  } catch (error) {
    const failedReport: ITestReport = {
      pass: false,
      testName: params.matcher,
      message: handleError(error),
    };

    if (!params.isDebug) {
      failedReport.trace = params.trace;
    }

    runtime.internalEvents.emit("test_end", failedReport);
    if (params.isDebug) {
      return failedReport;
    }
  }
}

function createMatcherFn(params: ICreateMatcherParam) {
  const {
    matcher,
    isNot,
    commandName,
    channelId,
    isDebug,
    isCascade,
    cordeBot,
    guildId,
    mustSendCommand,
  } = params;
  const { testCollector, configs } = runtime;

  if (!testCollector.currentTestFile?.isInsideTestClosure && !isCascade) {
    throw new Error("command can only be used inside a test(it) closure");
  }

  return (
    ...args: any[]
  ): Promise<ITestReport | void | ((...args: any[]) => Promise<ITestReport>)> => {
    const trace = getStackTrace(undefined, true, matcher);
    // If someone pass expect.any, we must invoke it to return
    // the Any matcher.

    // If the suite is already marked as failed,
    // There is no need to run other tests.
    // Same for command assertion
    if (testCollector.currentSuite?.markedAsFailed) {
      return Promise.reject();
    }

    args = args.map((arg) => {
      if (arg === any) {
        return arg();
      }
      return arg;
    });

    const matcherFn = pickFn(matcher as KeyOfMatcher);

    let bot = runtime.bot;

    if (isDebug) {
      bot = cordeBot ?? runtime.bot;
    }

    const props = new CommandState({
      isNot,
      cordeBot: bot,
      command: commandName,
      timeout: configs.commandTimeout,
      guildId: guildId ?? configs.guildId,
      channelId: channelId ?? configs.channelId,
      testName: matcher,
      isCascade: isCascade ?? false,
      mustSendCommand: isNullOrUndefined(mustSendCommand) ? true : !!mustSendCommand,
    });

    const fn = matcherFn.bind(props, ...args);
    params.trace = trace;
    return new CommandPromise(async (resolve, reject) => {
      try {
        const result = await resolveTestFunction(params, fn);
        resolve(result);
      } catch (error) {
        reject(error);
      }
    }, params);
  };
}

function handleError(error: any) {
  if (error instanceof Error) {
    return error.message + "\n" + error.stack;
  }
  return error;
}

function createLocalCommand(isDebug: boolean) {
  let localCommand: any = {};

  localCommand = (commandName: any, channelId?: string, cordeBot?: ICordeBot) => {
    const commandReturn: any = {
      should: {
        not: {},
      },
    };

    forEachPropsName(matchers, (matcher) => {
      commandReturn.should[matcher] = createMatcherFn({
        commandName,
        isDebug,
        isNot: false,
        matcher,
        channelId,
        cordeBot,
      });
      commandReturn.should.not[matcher] = createMatcherFn({
        commandName,
        isDebug,
        isNot: true,
        matcher,
        channelId,
        cordeBot,
      });
    });

    return commandReturn;
  };

  localCommand.any = any;

  return localCommand;
}

export const command = createLocalCommand(false) as corde.ICommand;

type DebugTypes<T> = {
  [P in keyof T]: T[P] extends (...args: any[]) => Promise<any>
    ? (...params: Parameters<T[P]>) => Promise<{
        pass: boolean;
        testName: string;
        message: string;
        trace?: string;
      }>
    : T[P] extends Record<string, any>
    ? DebugTypes<T[P]>
    : T[P];
};

type MatchersWithNot = {
  not: typeof matchers;
} & typeof matchers;

type Matchers = {
  should: MatchersWithNot;
};

interface DebugFn {
  <T>(value: T, channelId?: string, cordeBot?: ICordeBot): DebugTypes<Matchers>;
}

/**
 * Testing object for corde
 * This is not used in production
 *
 * @internal
 */
export const debugCommand = createLocalCommand(true) as DebugFn;
