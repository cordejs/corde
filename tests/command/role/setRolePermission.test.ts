import MockDiscord from "../../mocks/mockDiscord";
import { createCordeBotWithMockedFunctions, testHelper } from "../../testHelper";
import { ICordeBot, ITestReport } from "../../../src/types";
import { calcPermissionsValue } from "../../../src/utils/calcPermissionsValue";
import { MockEvents } from "../../mocks/mockEvents";
import { debugCommand } from "../../../src/command";
import { Permission } from "../../../src";

const testName = "setRolePermission";

const failReport: ITestReport = {
  pass: false,
  testName,
};

const passReport: ITestReport = {
  pass: true,
  testName,
};

let mockDiscord = new MockDiscord();
let cordeClient = createCordeBotWithMockedFunctions(mockDiscord, mockDiscord.client);

function debugCon(customCommand?: string, customChannelId?: string, customClient?: ICordeBot) {
  return debugCommand(customCommand ?? "con", customChannelId, customClient ?? cordeClient);
}

describe(`testing ${testName} function`, () => {
  beforeEach(() => {
    [mockDiscord, cordeClient] = testHelper.initCommandTestsFixtures();
  });

  it("should fail due to undefined roleIdentifier", async () => {
    const report = await debugCon()
      // @ts-ignore
      .should.setRolePermission(undefined, ["ADD_REACTIONS"]);
    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return false due to invalid permission parameter (object)", async () => {
    const report = await debugCon()
      // @ts-ignore
      .should.setRolePermission({ id: "123" }, {});
    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to invalid permission value", async () => {
    const report = await debugCon()
      // @ts-ignore
      .should.setRolePermission({ id: "123" }, ["BANANA"]);
    expect(report).toMatchObject(failReport);
  });

  it("should return false due to invalid permission parameter (undefined)", async () => {
    const report = await debugCon()
      // @ts-ignore
      .should.setRolePermission({ id: "123" }, {});

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return false due to not found role", async () => {
    cordeClient.findRole = jest.fn().mockReturnValue(null);
    const report = await debugCon().should.setRolePermission({ id: "123" }, "ATTACH_FILES");

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to role permissions was not updated", async () => {
    const report = await debugCon().should.setRolePermission({ id: "123" }, "ATTACH_FILES");
    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return true due to isNot true and no role change", async () => {
    const report = await debugCon().should.not.setRolePermission({ id: "123" }, "ATTACH_FILES");
    expect(report).toEqual(passReport);
    expect(report).toMatchSnapshot();
  });

  it("should return true due to role changed permissions (isNot false)", async () => {
    const mockEvent = new MockEvents(cordeClient, mockDiscord);
    const mockRole = mockDiscord.mockRole(
      "test role",
      // @ts-expect-error
      calcPermissionsValue(...mockDiscord.role.permissions.toArray().map((p) => Permission[p])),
    );

    mockEvent.mockOnceRolePermissionsUpdate(mockRole);
    const report = await debugCon().should.setRolePermission(
      { id: "123" },
      ...mockRole.permissions.toArray(),
    );
    expect(report).toEqual(passReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a failed test due to permissions should not change (isNot true)", async () => {
    const mockRole = mockDiscord.mockRole(
      "test role",
      // @ts-expect-error
      calcPermissionsValue(Permission.ADMINISTRATOR),
    );
    const mockEvent = new MockEvents(cordeClient, mockDiscord);

    mockEvent.mockOnceRolePermissionsUpdate(mockRole);
    const report = await debugCon().should.not.setRolePermission({ id: "123" }, "ADMINISTRATOR");
    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due to permissions should not change (isNot true)", async () => {
    const mockRole = mockDiscord.mockRole(
      "test role",
      // @ts-expect-error
      calcPermissionsValue(Permission.ADMINISTRATOR, Permission.BAN_MEMBERS),
    );
    const mockEvent = new MockEvents(cordeClient, mockDiscord);

    mockEvent.mockOnceRolePermissionsUpdate(mockRole);
    const report = await debugCon().should.not.setRolePermission(
      { id: "123" },
      "ADMINISTRATOR",
      "BAN_MEMBERS",
    );

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due to permissions should not change (isNot true)", async () => {
    const mockRole = mockDiscord.mockRole(
      "test role",
      // @ts-expect-error
      calcPermissionsValue(Permission.ADMINISTRATOR, Permission.BAN_MEMBERS, Permission.CONNECT),
    );

    const mockEvent = new MockEvents(cordeClient, mockDiscord);

    mockEvent.mockOnceRolePermissionsUpdate(mockRole);

    const report = await debugCon().should.not.setRolePermission(
      { id: "123" },
      "ADMINISTRATOR",
      "BAN_MEMBERS",
      "CONNECT",
    );

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due expected name did not match to received", async () => {
    const mockEvent = new MockEvents(cordeClient, mockDiscord);
    mockEvent.mockOnceRolePermissionsUpdate(mockDiscord.role);
    const report = await debugCon().should.setRolePermission({ id: "123" }, "ATTACH_FILES");

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a not passed test due expected name did not match to received", async () => {
    const mockEvent = new MockEvents(cordeClient, mockDiscord);
    mockEvent.mockOnceRolePermissionsUpdate(mockDiscord.role);
    const report = await debugCon()
      // @ts-ignore
      .should.setRolePermission({ id: "123" }, null);

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a failed test due to failure in message sending", async () => {
    cordeClient.findRole = jest.fn().mockReturnValue(mockDiscord.role);
    cordeClient.fetchRole = jest.fn().mockReturnValue(null);

    const errorMessage = "can not send message to channel x";
    cordeClient.sendTextMessage = jest
      .fn()
      .mockImplementation(() => Promise.reject(new Error(errorMessage)));

    const report = await debugCon().should.setRolePermission({ id: "123" }, "ATTACH_FILES");

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });
});
