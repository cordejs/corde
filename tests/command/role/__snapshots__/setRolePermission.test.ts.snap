// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing setRolePermission function should fail due to role permissions was not updated 1`] = `
Object {
  "message": "    
    expected: role permissions change to: ATTACH_FILES
    received: permissions were not changed
",
  "pass": false,
  "testName": "setRolePermission",
}
`;

exports[`testing setRolePermission function should fail due to undefined roleIdentifier 1`] = `
Object {
  "message": "    
    expected: data to identifier the role (id or name)
    received: null
",
  "pass": false,
  "testName": "setRolePermission",
}
`;

exports[`testing setRolePermission function should return a failed test due to failure in message sending 1`] = `
Object {
  "message": "    
    can not send message to channel x
",
  "pass": false,
  "testName": "setRolePermission",
}
`;

exports[`testing setRolePermission function should return a failed test due to permissions should not change (isNot true) 1`] = `
Object {
  "message": "    
    expected: role permissions not change to: ADMINISTRATOR
    received: ADMINISTRATOR (and 30 others)
",
  "pass": false,
  "testName": "setRolePermission",
}
`;

exports[`testing setRolePermission function should return a not passed test due expected name did not match to received 1`] = `
Object {
  "message": "    
    expected: role permissions change to: ATTACH_FILES
    received: 
",
  "pass": false,
  "testName": "setRolePermission",
}
`;

exports[`testing setRolePermission function should return a not passed test due expected name did not match to received 2`] = `
Object {
  "message": "    
    [32m- Expected[39m
    [31m+ Received[39m
    
    [2m  Array [[22m
    [32m-   \\"1\\",[39m
    [32m-   \\"2\\",[39m
    [32m-   \\"4\\",[39m
    [32m-   \\"8\\",[39m
    [32m-   \\"16\\",[39m
    [32m-   \\"32\\",[39m
    [32m-   \\"64\\",[39m
    [32m-   \\"128\\",[39m
    [32m-   \\"256\\",[39m
    [32m-   \\"512\\",[39m
    [32m-   \\"1024\\",[39m
    [32m-   \\"2048\\",[39m
    [32m-   \\"4096\\",[39m
    [32m-   \\"8192\\",[39m
    [32m-   \\"16384\\",[39m
    [32m-   \\"32768\\",[39m
    [32m-   \\"65536\\",[39m
    [32m-   \\"131072\\",[39m
    [32m-   \\"262144\\",[39m
    [32m-   \\"524288\\",[39m
    [32m-   \\"1048576\\",[39m
    [32m-   \\"2097152\\",[39m
    [32m-   \\"4194304\\",[39m
    [32m-   \\"8388608\\",[39m
    [32m-   \\"16777216\\",[39m
    [32m-   \\"33554432\\",[39m
    [32m-   \\"67108864\\",[39m
    [32m-   \\"134217728\\",[39m
    [32m-   \\"268435456\\",[39m
    [32m-   \\"536870912\\",[39m
    [32m-   \\"1073741824\\",[39m
    [32m-   \\"2147483648\\",[39m
    [32m-   \\"CREATE_INSTANT_INVITE\\",[39m
    [32m-   \\"KICK_MEMBERS\\",[39m
    [32m-   \\"BAN_MEMBERS\\",[39m
    [32m-   \\"ADMINISTRATOR\\",[39m
    [32m-   \\"MANAGE_CHANNELS\\",[39m
    [32m-   \\"MANAGE_GUILD\\",[39m
    [32m-   \\"ADD_REACTIONS\\",[39m
    [32m-   \\"VIEW_AUDIT_LOG\\",[39m
    [32m-   \\"VIEW_CHANNEL\\",[39m
    [32m-   \\"SEND_MESSAGES\\",[39m
    [32m-   \\"SEND_TTS_MESSAGES\\",[39m
    [32m-   \\"MANAGE_MESSAGES\\",[39m
    [32m-   \\"EMBED_LINKS\\",[39m
    [32m-   \\"ATTACH_FILES\\",[39m
    [32m-   \\"READ_MESSAGE_HISTORY\\",[39m
    [32m-   \\"MENTION_EVERYONE\\",[39m
    [32m-   \\"USE_EXTERNAL_EMOJIS\\",[39m
    [32m-   \\"CONNECT\\",[39m
    [32m-   \\"SPEAK\\",[39m
    [32m-   \\"MUTE_MEMBERS\\",[39m
    [32m-   \\"DEAFEN_MEMBERS\\",[39m
    [32m-   \\"MOVE_MEMBERS\\",[39m
    [32m-   \\"USE_VAD\\",[39m
    [32m-   \\"PRIORITY_SPEAKER\\",[39m
    [32m-   \\"CHANGE_NICKNAME\\",[39m
    [32m-   \\"MANAGE_NICKNAMES\\",[39m
    [32m-   \\"MANAGE_ROLES\\",[39m
    [32m-   \\"MANAGE_WEBHOOKS\\",[39m
    [32m-   \\"STREAM\\",[39m
    [32m-   \\"VIEW_GUILD_INSIGHTS\\",[39m
    [32m-   \\"MANAGE_EMOJIS_AND_STICKERS\\",[39m
    [32m-   \\"USE_APPLICATION_COMMANDS\\",[39m
    [32m-   \\"REQUEST_TO_SPEAK\\",[39m
    [32m-   \\"4294967296\\",[39m
    [32m-   \\"MANAGE_THREADS\\",[39m
    [32m-   \\"17179869184\\",[39m
    [32m-   \\"CREATE_PUBLIC_THREADS\\",[39m
    [32m-   \\"34359738368\\",[39m
    [32m-   \\"CREATE_PRIVATE_THREADS\\",[39m
    [32m-   \\"68719476736\\",[39m
    [32m-   \\"USE_EXTERNAL_STICKERS\\",[39m
    [32m-   \\"137438953472\\",[39m
    [32m-   \\"SEND_MESSAGES_IN_THREADS\\",[39m
    [32m-   \\"274877906944\\",[39m
    [32m-   \\"START_EMBEDDED_ACTIVITIES\\",[39m
    [32m-   \\"549755813888\\",[39m
    [32m-   \\"MODERATE_MEMBERS\\",[39m
    [32m-   \\"1099511627776\\",[39m
    [32m-   \\"MANAGE_EVENTS\\",[39m
    [32m-   \\"8589934592\\",[39m
    [32m-   \\"USE_PRIVATE_THREADS\\",[39m
    [32m-   \\"8589934593\\",[39m
    [32m-   \\"USE_PUBLIC_THREADS\\",[39m
    [32m-   \\"8589934594\\",[39m
    [31m+   null,[39m
    [2m  ][22m
",
  "pass": false,
  "testName": "setRolePermission",
}
`;

exports[`testing setRolePermission function should return a not passed test due to permissions should not change (isNot true) 1`] = `
Object {
  "message": "    
    expected: role permissions not change to: ADMINISTRATOR and BAN_MEMBERS
    received: ADMINISTRATOR (and 30 others)
",
  "pass": false,
  "testName": "setRolePermission",
}
`;

exports[`testing setRolePermission function should return a not passed test due to permissions should not change (isNot true) 2`] = `
Object {
  "message": "    
    expected: role permissions not change to: ADMINISTRATOR (and 2 others)
    received: ADMINISTRATOR (and 30 others)
",
  "pass": false,
  "testName": "setRolePermission",
}
`;

exports[`testing setRolePermission function should return false due to invalid permission parameter (object) 1`] = `
Object {
  "message": "    
    [32m- Expected[39m
    [31m+ Received[39m
    
    [2m  Array [[22m
    [32m-   \\"1\\",[39m
    [32m-   \\"2\\",[39m
    [32m-   \\"4\\",[39m
    [32m-   \\"8\\",[39m
    [32m-   \\"16\\",[39m
    [32m-   \\"32\\",[39m
    [32m-   \\"64\\",[39m
    [32m-   \\"128\\",[39m
    [32m-   \\"256\\",[39m
    [32m-   \\"512\\",[39m
    [32m-   \\"1024\\",[39m
    [32m-   \\"2048\\",[39m
    [32m-   \\"4096\\",[39m
    [32m-   \\"8192\\",[39m
    [32m-   \\"16384\\",[39m
    [32m-   \\"32768\\",[39m
    [32m-   \\"65536\\",[39m
    [32m-   \\"131072\\",[39m
    [32m-   \\"262144\\",[39m
    [32m-   \\"524288\\",[39m
    [32m-   \\"1048576\\",[39m
    [32m-   \\"2097152\\",[39m
    [32m-   \\"4194304\\",[39m
    [32m-   \\"8388608\\",[39m
    [32m-   \\"16777216\\",[39m
    [32m-   \\"33554432\\",[39m
    [32m-   \\"67108864\\",[39m
    [32m-   \\"134217728\\",[39m
    [32m-   \\"268435456\\",[39m
    [32m-   \\"536870912\\",[39m
    [32m-   \\"1073741824\\",[39m
    [32m-   \\"2147483648\\",[39m
    [32m-   \\"CREATE_INSTANT_INVITE\\",[39m
    [32m-   \\"KICK_MEMBERS\\",[39m
    [32m-   \\"BAN_MEMBERS\\",[39m
    [32m-   \\"ADMINISTRATOR\\",[39m
    [32m-   \\"MANAGE_CHANNELS\\",[39m
    [32m-   \\"MANAGE_GUILD\\",[39m
    [32m-   \\"ADD_REACTIONS\\",[39m
    [32m-   \\"VIEW_AUDIT_LOG\\",[39m
    [32m-   \\"VIEW_CHANNEL\\",[39m
    [32m-   \\"SEND_MESSAGES\\",[39m
    [32m-   \\"SEND_TTS_MESSAGES\\",[39m
    [32m-   \\"MANAGE_MESSAGES\\",[39m
    [32m-   \\"EMBED_LINKS\\",[39m
    [32m-   \\"ATTACH_FILES\\",[39m
    [32m-   \\"READ_MESSAGE_HISTORY\\",[39m
    [32m-   \\"MENTION_EVERYONE\\",[39m
    [32m-   \\"USE_EXTERNAL_EMOJIS\\",[39m
    [32m-   \\"CONNECT\\",[39m
    [32m-   \\"SPEAK\\",[39m
    [32m-   \\"MUTE_MEMBERS\\",[39m
    [32m-   \\"DEAFEN_MEMBERS\\",[39m
    [32m-   \\"MOVE_MEMBERS\\",[39m
    [32m-   \\"USE_VAD\\",[39m
    [32m-   \\"PRIORITY_SPEAKER\\",[39m
    [32m-   \\"CHANGE_NICKNAME\\",[39m
    [32m-   \\"MANAGE_NICKNAMES\\",[39m
    [32m-   \\"MANAGE_ROLES\\",[39m
    [32m-   \\"MANAGE_WEBHOOKS\\",[39m
    [32m-   \\"STREAM\\",[39m
    [32m-   \\"VIEW_GUILD_INSIGHTS\\",[39m
    [32m-   \\"MANAGE_EMOJIS_AND_STICKERS\\",[39m
    [32m-   \\"USE_APPLICATION_COMMANDS\\",[39m
    [32m-   \\"REQUEST_TO_SPEAK\\",[39m
    [32m-   \\"4294967296\\",[39m
    [32m-   \\"MANAGE_THREADS\\",[39m
    [32m-   \\"17179869184\\",[39m
    [32m-   \\"CREATE_PUBLIC_THREADS\\",[39m
    [32m-   \\"34359738368\\",[39m
    [32m-   \\"CREATE_PRIVATE_THREADS\\",[39m
    [32m-   \\"68719476736\\",[39m
    [32m-   \\"USE_EXTERNAL_STICKERS\\",[39m
    [32m-   \\"137438953472\\",[39m
    [32m-   \\"SEND_MESSAGES_IN_THREADS\\",[39m
    [32m-   \\"274877906944\\",[39m
    [32m-   \\"START_EMBEDDED_ACTIVITIES\\",[39m
    [32m-   \\"549755813888\\",[39m
    [32m-   \\"MODERATE_MEMBERS\\",[39m
    [32m-   \\"1099511627776\\",[39m
    [32m-   \\"MANAGE_EVENTS\\",[39m
    [32m-   \\"8589934592\\",[39m
    [32m-   \\"USE_PRIVATE_THREADS\\",[39m
    [32m-   \\"8589934593\\",[39m
    [32m-   \\"USE_PUBLIC_THREADS\\",[39m
    [32m-   \\"8589934594\\",[39m
    [31m+   Object {},[39m
    [2m  ][22m
",
  "pass": false,
  "testName": "setRolePermission",
}
`;

exports[`testing setRolePermission function should return false due to invalid permission parameter (undefined) 1`] = `
Object {
  "message": "    
    [32m- Expected[39m
    [31m+ Received[39m
    
    [2m  Array [[22m
    [32m-   \\"1\\",[39m
    [32m-   \\"2\\",[39m
    [32m-   \\"4\\",[39m
    [32m-   \\"8\\",[39m
    [32m-   \\"16\\",[39m
    [32m-   \\"32\\",[39m
    [32m-   \\"64\\",[39m
    [32m-   \\"128\\",[39m
    [32m-   \\"256\\",[39m
    [32m-   \\"512\\",[39m
    [32m-   \\"1024\\",[39m
    [32m-   \\"2048\\",[39m
    [32m-   \\"4096\\",[39m
    [32m-   \\"8192\\",[39m
    [32m-   \\"16384\\",[39m
    [32m-   \\"32768\\",[39m
    [32m-   \\"65536\\",[39m
    [32m-   \\"131072\\",[39m
    [32m-   \\"262144\\",[39m
    [32m-   \\"524288\\",[39m
    [32m-   \\"1048576\\",[39m
    [32m-   \\"2097152\\",[39m
    [32m-   \\"4194304\\",[39m
    [32m-   \\"8388608\\",[39m
    [32m-   \\"16777216\\",[39m
    [32m-   \\"33554432\\",[39m
    [32m-   \\"67108864\\",[39m
    [32m-   \\"134217728\\",[39m
    [32m-   \\"268435456\\",[39m
    [32m-   \\"536870912\\",[39m
    [32m-   \\"1073741824\\",[39m
    [32m-   \\"2147483648\\",[39m
    [32m-   \\"CREATE_INSTANT_INVITE\\",[39m
    [32m-   \\"KICK_MEMBERS\\",[39m
    [32m-   \\"BAN_MEMBERS\\",[39m
    [32m-   \\"ADMINISTRATOR\\",[39m
    [32m-   \\"MANAGE_CHANNELS\\",[39m
    [32m-   \\"MANAGE_GUILD\\",[39m
    [32m-   \\"ADD_REACTIONS\\",[39m
    [32m-   \\"VIEW_AUDIT_LOG\\",[39m
    [32m-   \\"VIEW_CHANNEL\\",[39m
    [32m-   \\"SEND_MESSAGES\\",[39m
    [32m-   \\"SEND_TTS_MESSAGES\\",[39m
    [32m-   \\"MANAGE_MESSAGES\\",[39m
    [32m-   \\"EMBED_LINKS\\",[39m
    [32m-   \\"ATTACH_FILES\\",[39m
    [32m-   \\"READ_MESSAGE_HISTORY\\",[39m
    [32m-   \\"MENTION_EVERYONE\\",[39m
    [32m-   \\"USE_EXTERNAL_EMOJIS\\",[39m
    [32m-   \\"CONNECT\\",[39m
    [32m-   \\"SPEAK\\",[39m
    [32m-   \\"MUTE_MEMBERS\\",[39m
    [32m-   \\"DEAFEN_MEMBERS\\",[39m
    [32m-   \\"MOVE_MEMBERS\\",[39m
    [32m-   \\"USE_VAD\\",[39m
    [32m-   \\"PRIORITY_SPEAKER\\",[39m
    [32m-   \\"CHANGE_NICKNAME\\",[39m
    [32m-   \\"MANAGE_NICKNAMES\\",[39m
    [32m-   \\"MANAGE_ROLES\\",[39m
    [32m-   \\"MANAGE_WEBHOOKS\\",[39m
    [32m-   \\"STREAM\\",[39m
    [32m-   \\"VIEW_GUILD_INSIGHTS\\",[39m
    [32m-   \\"MANAGE_EMOJIS_AND_STICKERS\\",[39m
    [32m-   \\"USE_APPLICATION_COMMANDS\\",[39m
    [32m-   \\"REQUEST_TO_SPEAK\\",[39m
    [32m-   \\"4294967296\\",[39m
    [32m-   \\"MANAGE_THREADS\\",[39m
    [32m-   \\"17179869184\\",[39m
    [32m-   \\"CREATE_PUBLIC_THREADS\\",[39m
    [32m-   \\"34359738368\\",[39m
    [32m-   \\"CREATE_PRIVATE_THREADS\\",[39m
    [32m-   \\"68719476736\\",[39m
    [32m-   \\"USE_EXTERNAL_STICKERS\\",[39m
    [32m-   \\"137438953472\\",[39m
    [32m-   \\"SEND_MESSAGES_IN_THREADS\\",[39m
    [32m-   \\"274877906944\\",[39m
    [32m-   \\"START_EMBEDDED_ACTIVITIES\\",[39m
    [32m-   \\"549755813888\\",[39m
    [32m-   \\"MODERATE_MEMBERS\\",[39m
    [32m-   \\"1099511627776\\",[39m
    [32m-   \\"MANAGE_EVENTS\\",[39m
    [32m-   \\"8589934592\\",[39m
    [32m-   \\"USE_PRIVATE_THREADS\\",[39m
    [32m-   \\"8589934593\\",[39m
    [32m-   \\"USE_PUBLIC_THREADS\\",[39m
    [32m-   \\"8589934594\\",[39m
    [31m+   Object {},[39m
    [2m  ][22m
",
  "pass": false,
  "testName": "setRolePermission",
}
`;

exports[`testing setRolePermission function should return false due to not found role 1`] = `
Object {
  "message": "    
    expected: role with id 123
    received: null
",
  "pass": false,
  "testName": "setRolePermission",
}
`;

exports[`testing setRolePermission function should return true due to isNot true and no role change 1`] = `
Object {
  "pass": true,
  "testName": "setRolePermission",
}
`;

exports[`testing setRolePermission function should return true due to role changed permissions (isNot false) 1`] = `
Object {
  "pass": true,
  "testName": "setRolePermission",
}
`;
